DEV-43: Update use_cases.md with MVP details (#38)
dcsil/lexislabs: product_research/use_cases.md

### User interaction

> The user is shown a button to begin mic input and instructions to read aloud a displayed text in their target learning language (e.g. a short story). The user will be able to select a sentence to read. There will also be a progress bar for the number of sentences read, out of the total number in the passage. Upon pressing the button and completing the reading of the selected sentence, pronunciation feedback will be displayed by highlighting mispronounced words, along with providing a button to view the aggregated summary based on the feedback on the latest attempts of each sentence, from which the user will also be able to navigate to the feedback on older attempts. The user can click on feedback items for more detailed or graphical explanations of their pronunciation errors. This will include the option to listen to an ideal pronounciation of the mispronounced words. There will be the option to play back the user's reading attempt, and errors will be highlighted on the timeline for the playback.

Author: removed

> There will be a screen for importing text, either by copy-pasting it into a text field or from a file. The user will also be able to view their reading karaoke history, which will be searchable using various metrics. The user will be able to view the detailed feedback they received for any attempt in their history. Also, the history will be tied to their account, for which login, password management, and account deletion will be available.

Reviewer: ```suggestion
There will be a screen for importing text, either by copy-pasting it into a text field or from a file. The user will also be able to view their reading karaoke history, which will be searchable using various metrics. The user will be able to view the detailed feedback they received for any attempt in their history. Additionally, users' history will be linked to their accounts, allowing for login, password management, and account deletion.
```

The UI will be available in multiple languages.

### Technical implementation

> The user will have the texts they have imported or generated stored in the Postgres DB, so if they aren't already cached in the frontend, the backend can send them over.

Reviewer: ```suggestion
The texts that users import or generate will be stored in the Postgres DB. If these texts are not already cached on the frontend, they will be retrieved from the backend as needed.
```

Upon finishing mic input, the user's recording is sent to the backend, which is sent to the Azure Speech AI Pronunciation Assessment API. The reference text is also sent to Azure. As feedback is returned from Azure, the backend will forward the updated scores and feedback for the frontend to display. Ideal pronunciation of each mispronounced word can be retrieved from the Forvo API. For audio playback, the recording can be temporarily stored on the frontend.

> Text importing involves sending the copy-pasted or imported text to the backend to te stored in the DB. The history would be a table of the relevant feedback details for each attempt, though it would be cached in the frontend for searching.

Reviewer: ```suggestion
Text importing will involve sending the copy-pasted or imported text to the backend to be stored in the database. The history will be maintained as a table of relevant feedback details for each attempt, and it will be cached on the frontend for easier searching.
```

  <em>Figure 1: Reading karaoke UI sketch</em>
</p>

### MVP

- Importing copy-pasted text
- Viewing aggregated latest feedback on all sentences in a given passage, with a way to navigate to older feedback
- Button to start/stop recording
- Progress bar for number of sentences read out of total in passage
- Can select sentence from passage to practice
- Detailed feedback available for each error, which would be a highlighted word

> - Ideal pronounciation available for each error

Reviewer 2: As it's currently written, it's unclear what form this "ideal" will take. I think we agreed that it should be a playable audio clip (as opposed to, for example, an IPA transcription).

  <img src="./assets/Dashboard.png" />
  <br>
  <em>Figure 4: Progress dashboard UI sketch</em>
</p>

### MVP

- Shows learning streak, defined as the number of consecutive days (dependent on the user's local time zone) in which the user completed at least one activity
- Shows the total number of activities completed in the last 30 days
- Has a "continue where you left off" button

## 5. AI-powered reading passage generation

### User interaction

The user is presented with sliders and options describing various aspects of the desired text, along with a switch to indicate whether to allow AI-generated text or to require text sourced from the internet. The user can then choose to generate the text, review the result, and choose whether to use it or regenerate with different options.

### Technical implementation

Implementing the options is a matter of well-designed prompts to be given to the Cohere Aya LLM by the backend, while the AI-generated versus internet-sourced switch concerns whether or not to use RAG to get the text. The backend will also need to do some sanitization on the generated text. Then it can send it to the frontend, which will present the text for review.

> ### MVP

Reviewer 3: I think we need a UI sketch here as well. I sent it in our group chat.

- Has a slider for vocabulary difficulty

> - Has a switch for whether to use RAG

Reviewer 3: Follow up on the conversation we are having about whether or not to include RAG in our MVP.

Author: removed

There will also be a "continue where you left off" button to jump to where the user stopped in their previous session.

### Technical implementation

The backend requests the necessary user data from the DB, and generates a data structure that contains all the statistics to be displayed. The frontend parses the data structure to populate graphical elements. The frontend should cache the report to reduce demand on BE resources if the user frequently opens the dashboard page.

Considering that the history of activities is stored, the most recent of those can be queried to determine where to jump to when "continue where you left off" is pressed, and this can also be cached by the frontend.

### UI sketch

<p align="center">
  <img src="./assets/Dashboard.png" />
  <br>
  <em>Figure 4: Progress dashboard UI sketch</em>
</p>

### MVP

- Shows learning streak, defined as the number of consecutive days (dependent on the user's local time zone) in which the user completed at least one activity
- Shows the total number of activities completed in the last 30 days

> - Shows an error breakdown chart containing total counts of each type of error

Reviewer 2: The breakdown chart should not be in the MVP.
